### PR TITLE
Fix code Hash

### PR DESCRIPTION
**Pull request by karthikeyan-cere**
Fetch code hash from events

### CODE REVIEW

File: scripts/tokenomics-emulation/config.json [hunk 1]
```diff
@@ -113,7 +113,8 @@
       {
         "name": "send-ddc-transaction",
         "amount": 1,
-        "sendData": "This is test data"
+        "string_lenght": 10,
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/config.json:116]: We need bytes. It should be 128 by default.

File: scripts/tokenomics-emulation/network.ts [hunk 2]
```diff
@@ -247,7 +247,7 @@ class Network {
     } else if (status.isFinalized) {
       const hash = status.asFinalized.toHex();
       console.info(`Transaction has been included in blockHash ${hash}`);
-      if (handleEvents !== undefined) {
+      if (typeof handleEvents !== 'undefined') {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/network.ts:250]: Why do we need this?

Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/network.ts:250]: To check the type of variable.

File: scripts/tokenomics-emulation/config.json (hunk 1 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/config.json]: Updated

File: scripts/tokenomics-emulation/network.ts (hunk 2 above)
Comment (AndreiNavoichyk) [reply scripts/tokenomics-emulation/network.ts:250]: What is wrong with the previous approach?

File: scripts/tokenomics-emulation/ddc-smart-contract.ts [hunk 3]
```diff
@@ -142,7 +142,24 @@ class DdcSmartContract {
...
-  }
+   }
+  
+   /**
+   * Process the events to set code hash
+   * @param events Events
+   */
+    private async handleEventsForCodeHash(events) {
+      console.log("Handling Events");
+      events.forEach((event) => {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/ddc-smart-contract.ts:153]: Do not copy-paste handlers between SCs. Move to a separate class and reuse in both.

File: scripts/tokenomics-emulation/network.ts (hunk 2 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/network.ts]: reverted to previous state

File: scripts/tokenomics-emulation/ddc-smart-contract.ts (hunk 3 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/ddc-smart-contract.ts]: Done. Created a new EventHandler class and handling there

File: scripts/tokenomics-emulation/event-handlers.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+class EventHandlers {
+
+
+  /**
+   * Process the events to set smart contract address
+   * @param events Events
+   */
+   public static async handleEventsForSmartContractAddress(events, param) {
+     console.log("Handling Events");
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/event-handlers.ts:12]: Add more specific log so it is clear which params you are handling

File: scripts/tokenomics-emulation/event-handlers.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import fs from "fs";
+
+class EventHandlers {
+
+
+  /**
+   * Process the events to set smart contract address
+   * @param events Events
+   */
+   public static async handleEventsForSmartContractAddress(events, param) {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/event-handlers.ts:11]: What is the `param`? Rename it.

File: scripts/tokenomics-emulation/event-handlers.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+       
+      }
+    });
+   }
+  
+   /**
+   * Process the events to set code hash
+   * @param events Events
+   */
+    public static async handleEventsForCodeHash(events, param) {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/event-handlers.ts:30]: What is the param? Rename it.

File: scripts/tokenomics-emulation/event-handlers.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      console.log("Handling Events");
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/event-handlers.ts:31]: Add more specific log so it is clear which params you are handling

File: scripts/tokenomics-emulation/event-handlers.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      events.forEach((event) => {
+        const data = event.event.data[0].toString();
+        if (data.startsWith('0x')) {
+          configFile.emulations.sequence.find(sequence => sequence.name === param).code_hash = data;
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/event-handlers.ts:35]: Fix formatting

File: scripts/tokenomics-emulation/network.ts [hunk 9]
```diff
@@ -226,6 +226,7 @@ class Network {
     res,
     rej,
     handleEvents,
+    params,
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/network.ts:229]: No, it breaks other executions. You can pass params on the execution level.

File: scripts/tokenomics-emulation/event-handlers.ts (hunk 4 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/event-handlers.ts]: Added

File: scripts/tokenomics-emulation/event-handlers.ts (hunk 6 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/event-handlers.ts]: Renamed

File: scripts/tokenomics-emulation/event-handlers.ts (hunk 8 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/event-handlers.ts]: Formatted

File: scripts/tokenomics-emulation/event-handlers.ts (hunk 7 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/event-handlers.ts]: Added

File: scripts/tokenomics-emulation/event-handlers.ts (hunk 5 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/event-handlers.ts]: Renamed

File: scripts/tokenomics-emulation/network.ts (hunk 9 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/network.ts]: Done

File: scripts/tokenomics-emulation/cere-smart-contract.ts [hunk 10]
```diff
@@ -79,15 +80,17 @@ class CereSmartContract {
...
+  public async deploy(sender: KeyringPair, emulationName: string) {
+    console.log(`Deploy cere code to get code hash`);
     const code = new CodePromise(this.api, cere02Abi, cere01Wasm);
 
     const tx = await code.createBlueprint();
     return new Promise((res, rej) => {
       tx.signAndSend(
         sender,
-        Network.sendStatusCb.bind(this, res, rej, undefined)
+        Network.sendStatusCb.bind(this, res, rej, async (events) => {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/cere-smart-contract.ts:91]: Why is it `async`?

File: scripts/tokenomics-emulation/cere-smart-contract.ts [hunk 11]
```diff
@@ -118,26 +122,16 @@ class CereSmartContract {
       unsub
         .signAndSend(
           sender,
-          Network.sendStatusCb.bind(this, res, rej, this.handleEvents)
+          Network.sendStatusCb.bind(this, res, rej, async (events) => {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/cere-smart-contract.ts:125]: Why is it `async`?

File: scripts/tokenomics-emulation/event-handlers.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import configFile from "./config.json";
+import fs from "fs";
+
+class EventHandlers {
+  /**
+   * Process the events to set smart contract address
+   * @param events Events
+   */
+  public static async handleEventsForSmartContractAddress(events, smartContractKey: string) {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/event-handlers.ts:9]: Why is it async?

File: scripts/tokenomics-emulation/network.ts [hunk 13]
```diff
@@ -243,8 +243,8 @@ class Network {
     } else if (status.isFinalized) {
       const hash = status.asFinalized.toHex();
       console.info(`Transaction has been included in blockHash ${hash}`);
-      if (handleEvents !== undefined) {
-        handleEvents(events);
+      if (handleEvents !== 'undefined') {
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/network.ts:246]: Why do you need quotes here?

File: scripts/tokenomics-emulation/cere-smart-contract.ts (hunk 10 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts:91]: Removed

File: scripts/tokenomics-emulation/cere-smart-contract.ts (hunk 11 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/cere-smart-contract.ts:125]: Removed

File: scripts/tokenomics-emulation/event-handlers.ts (hunk 12 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/event-handlers.ts:9]: Removed

File: scripts/tokenomics-emulation/network.ts (hunk 13 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/network.ts:246]: Added while formating, removed

**Review by AndreiNavoichyk — APPROVED** [review]

**Pull request merged**